
Add fix for CAN-2004-0885, SSLCipherSuite bypass.  

Upstream-2.0: pending
Upstream-HEAD: committed
Upstream-Status: mod_ssl fix for CAN-2004-0885

--- httpd-2.0.52/modules/ssl/ssl_engine_kernel.c.can0885
+++ httpd-2.0.52/modules/ssl/ssl_engine_kernel.c
@@ -719,6 +719,21 @@
                 X509_free(peercert);
             }
         }
+        
+        /*
+         * Also check that SSLCipherSuite has been enforced as expected.
+         */
+        if (cipher_list) {
+            cipher = SSL_get_current_cipher(ssl);
+            if (sk_SSL_CIPHER_find(cipher_list, cipher) < 0) {
+                ap_log_rerror(APLOG_MARK, APLOG_ERR, 0, r,
+                             "SSL cipher suite not renegotiated: "
+                             "access to %s denied using cipher %s",
+                              r->filename,
+                              SSL_CIPHER_get_name(cipher));
+                return HTTP_FORBIDDEN;
+            }
+        }
     }
 
     /*
--- httpd-2.0.52/modules/ssl/ssl_engine_init.c.can0885
+++ httpd-2.0.52/modules/ssl/ssl_engine_init.c
@@ -439,6 +439,14 @@
      * Configure additional context ingredients
      */
     SSL_CTX_set_options(ctx, SSL_OP_SINGLE_DH_USE);
+
+#ifdef SSL_OP_NO_SESSION_RESUMPTION_ON_RENEGOTIATION
+    /* 
+     * Disallow a session from being resumed during a renegotiation,
+     * so that an acceptable cipher suite can be negotiated.
+     */
+    SSL_CTX_set_options(ctx, SSL_OP_NO_SESSION_RESUMPTION_ON_RENEGOTIATION);
+#endif
 }
 
 static void ssl_init_ctx_session_cache(server_rec *s,
